File: cpp/23_reverse_linked_list.cpp
/**
 * LeetCode #206: 反转链表
 * 题目: 反转单链表，返回新链表头。
 * 解法: 头插法 / 双指针迭代 - 遍历原链表，将每个节点以头插法插入新链表。
 *       prev=None 为新链表头，cur 从 head 遍历，每次头插。
 * 时间 O(n)，空间 O(1)。
 */

#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode *prev = nullptr, *cur = head;
        while (cur) {
            ListNode* nxt = cur->next;
            cur->next = prev;
            prev = cur;
            cur = nxt;
        }
        return prev;
    }
};

int main() {
    Solution sol;
    ListNode *n1 = new ListNode(1), *n2 = new ListNode(2),
             *n3 = new ListNode(3), *n4 = new ListNode(4), *n5 = new ListNode(5);
    n1->next = n2; n2->next = n3; n3->next = n4; n4->next = n5;
    ListNode* rev = sol.reverseList(n1);
    while (rev) { cout << rev->val << " "; rev = rev->next; } // 5 4 3 2 1
    cout << endl;
    return 0;
}

File: python/23_reverse_linked_list.py
#!/usr/bin/env python3
"""
LeetCode #206: 反转链表
题目: 给定单链表，反转后返回新链表的头节点。
解法: 头插法 / 双指针迭代 - 遍历原链表，将每个节点以头插法方式插入新链表。
      也可用递归。时间 O(n)，空间 O(1)。
"""


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


def reverseList(head: ListNode) -> ListNode:
    """
    头插法: prev=None 为新链表头，cur 从 head 开始遍历，
    每次将 cur.next 指向 prev，然后 prev=cur, cur=next。
    """
    prev = None
    cur = head
    while cur:
        nxt = cur.next
        cur.next = prev
        prev = cur
        cur = nxt
    return prev


def reverseList_recursive(head: ListNode) -> ListNode:
    """递归版"""
    if not head or not head.next:
        return head
    new_head = reverseList_recursive(head.next)
    head.next.next = head
    head.next = None
    return new_head


if __name__ == "__main__":
    # 1->2->3->4->5
    nodes = [ListNode(i) for i in range(1, 6)]
    for i in range(4): nodes[i].next = nodes[i+1]
    rev = reverseList(nodes[0])
    vals = []
    while rev:
        vals.append(rev.val)
        rev = rev.next
    print(vals)  # [5, 4, 3, 2, 1]
